homeassistant/components/unifi/sensor.py

```diff
@@ -3,183 +3,278 @@
+    """Create device registry entry for client."""
+    client = api.clients[obj_id]
+    return DeviceInfo(
+        connections={(CONNECTION_NETWORK_MAC, obj_id)},
+        default_manufacturer=client.oui,
+        default_name=client.name or client.hostname,
+    )
+
+
+@dataclass
+class UnifiEntityLoader(Generic[_HandlerT, _DataT]):
+    """Validate and load entities from different UniFi handlers."""
+
+    allowed_fn: Callable[[UniFiController, str], bool]
+    api_handler_fn: Callable[[aiounifi.Controller], _HandlerT]
+    available_fn: Callable[[UniFiController, str], bool]
+    device_info_fn: Callable[[aiounifi.Controller, str], DeviceInfo]
+    name_fn: Callable[[_DataT], str | None]
+    object_fn: Callable[[aiounifi.Controller, str], _DataT]
+    supported_fn: Callable[[UniFiController, str], bool | None]
+    unique_id_fn: Callable[[str], str]
+    value_fn: Callable[[UniFiController, _DataT], datetime | float]
+
+
+@dataclass
+class UnifiEntityDescription(
+    SensorEntityDescription, UnifiEntityLoader[_HandlerT, _DataT]
+):
+    """Class describing UniFi sensor entity."""
+
+
+ENTITY_DESCRIPTIONS: tuple[UnifiEntityDescription, ...] = (
+    UnifiEntityDescription[Clients, Client](
+        key="Bandwidth sensor RX",
+        entity_category=EntityCategory.DIAGNOSTIC,
+        native_unit_of_measurement=DATA_MEGABYTES,
+        has_entity_name=True,
+        allowed_fn=lambda controller, _: controller.option_allow_bandwidth_sensors,
+        api_handler_fn=lambda api: api.clients,
+        available_fn=lambda controller, _: controller.available,
+        device_info_fn=async_client_device_info_fn,
+        name_fn=lambda _: "RX",
+        object_fn=lambda api, obj_id: api.clients[obj_id],
+        supported_fn=lambda controller, _: controller.option_allow_bandwidth_sensors,
+        unique_id_fn=lambda obj_id: f"rx-{obj_id}",
+        value_fn=async_client_rx_value_fn,
+    ),
+    UnifiEntityDescription[Clients, Client](
+        key="Bandwidth sensor TX",
+        entity_category=EntityCategory.DIAGNOSTIC,
+        native_unit_of_measurement=DATA_MEGABYTES,
+        has_entity_name=True,
+        allowed_fn=lambda controller, _: controller.option_allow_bandwidth_sensors,
+        api_handler_fn=lambda api: api.clients,
+        available_fn=lambda controller, _: controller.available,
+        device_info_fn=async_client_device_info_fn,
+        name_fn=lambda _: "TX",
+        object_fn=lambda api, obj_id: api.clients[obj_id],
+        supported_fn=lambda controller, _: controller.option_allow_bandwidth_sensors,
+        unique_id_fn=lambda obj_id: f"tx-{obj_id}",
+        value_fn=async_client_tx_value_fn,
+    ),
+    UnifiEntityDescription[Clients, Client](
+        key="Uptime sensor",
+        device_class=SensorDeviceClass.TIMESTAMP,
+        entity_category=EntityCategory.DIAGNOSTIC,
+        has_entity_name=True,
+        allowed_fn=lambda controller, _: controller.option_allow_uptime_sensors,
```

Comment:
Why specify a second parameter if we never use it?

Reply:
This is normalised over the different platforms as it will be consolidated to one base definition down the line

https://github.com/home-assistant/core/blob/3d4ee5906da9f0537fff68c5c01bfce5c0bde771/homeassistant/components/unifi/switch.py#L200